lib/elastomer/client/docs.rb

```diff
@@ -534,10 +531,23 @@ def from_document( document )
         opts = {:body => document}
 
         if document.is_a? Hash
-          SPECIAL_KEYS_HASH.each do |key, field|
+          client.version_support.indexing_directives.each do |key, field|
             opts[key] = document.delete field if document.key? field
             opts[key] = document.delete field.to_sym if document.key? field.to_sym
           end
+
+          # COMPATIBILITY
+          # Fail fast if a consumer is attempting to use an indexing parameter
+          # for a different version of Elasticsearch. Elasticsearch 5+ is strict
+          # about parameter names so we need to either ignore these (in which
+          # case they would be indexed with the document) or fail-fast.
+          # Elasticsearch 2.X will happily ignore unknown parameters, but we
+          # felt it was best to consistently fail fast.
+          client.version_support.unsupported_indexing_directives.each do |key, field|
+            if document.key?(field) || document.key?(field.to_sym)
+              raise IncompatibleVersionException, "Elasticsearch #{client.version} does not support the '#{key}' indexing parameter"
```

Comment:
My thought here is that the `IncompatibleVersionException` should be used for features that are not available across both major versions of Elasticsearch that we support. For parameters, we should create a new `IllegalArgumentException` that is raised when an incompatible parameter is used. This mirrors the exception returned by Elasticsearch itself, and we will be using this exception across more and more the client library.

The distinction helps us know when a feature is just gone versus when a request needs a little tweaking.

Reply:
Yeah, I agree. I'll create a new exception for this.

Reply:
this makes sense :+1: